<a href="https://colab.research.google.com/github/dantrainor9/Cervical_cancer_risk_factors/blob/main/Project_2_Part_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [19]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn.preprocessing import StandardScaler
from sklearn.impute import SimpleImputer
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.pipeline import make_pipeline
from xgboost import XGBClassifier
from sklearn.ensemble import RandomForestClassifier
from lightgbm import LGBMClassifier
from sklearn.metrics import accuracy_score, classification_report

In [20]:
#risk factors affecting cervical cancer
path = '/content/drive/MyDrive/CodingDojo Data Science Bootcamp/07 Week 7 Logistic Regression and Regularization/risk_factors_cervical_cancer.csv'
df = pd.read_csv(path)
df.head()

Age Number of sexual partners  ... Citology Biopsy
0   18                       4.0  ...        0      0
1   15                       1.0  ...        0      0
2   34                       1.0  ...        0      0
3   52                       5.0  ...        0      0
4   46                       3.0  ...        0      0

[5 rows x 36 columns]

In [21]:
df.shape

(858, 36)

Multiple targets are present in the above dataset gathered from the UCI machine learning repository, as multiple diagnostic tools were used. However, as 'Biopsy' is the most common way of diagnosing cancer in the United States, I will be using that as my primary target after comparing the performance of the different diagnostic tests compared to the actual diagnosis (Dx:Cancer) 
This is a classification problem with 32 features.
There are 858 rows of data. 
I anticipate problems with missing data, as participants were not required to answer every question due to the sensitive nature of the survey.

In [22]:
df = df.replace('?', np.nan)

In [23]:
df.isna().sum()

Age                                     0
Number of sexual partners              26
First sexual intercourse                7
Num of pregnancies                     56
Smokes                                 13
Smokes (years)                         13
Smokes (packs/year)                    13
Hormonal Contraceptives               108
Hormonal Contraceptives (years)       108
IUD                                   117
IUD (years)                           117
STDs                                  105
STDs (number)                         105
STDs:condylomatosis                   105
STDs:cervical condylomatosis          105
STDs:vaginal condylomatosis           105
STDs:vulvo-perineal condylomatosis    105
STDs:syphilis                         105
STDs:pelvic inflammatory disease      105
STDs:genital herpes                   105
STDs:molluscum contagiosum            105
STDs:AIDS                             105
STDs:HIV                              105
STDs:Hepatitis B                  

In [24]:
X = df[['Age', 
        'Number of sexual partners', 
        'First sexual intercourse', 
        'Num of pregnancies', 
        'Smokes', 
        'Smokes (years)', 
        'Smokes (packs/year)', 
        'Hormonal Contraceptives', 
        'Hormonal Contraceptives (years)', 
        'IUD', 
        'IUD (years)', 
        'STDs', 
        'STDs (number)', 
        'STDs:condylomatosis', 
        'STDs:cervical condylomatosis', 
        'STDs:vaginal condylomatosis', 
        'STDs:vulvo-perineal condylomatosis', 
        'STDs:syphilis', 
        'STDs:pelvic inflammatory disease', 
        'STDs:genital herpes', 
        'STDs:molluscum contagiosum', 
        'STDs:AIDS', 
        'STDs:HIV', 
        'STDs:Hepatitis B', 
        'STDs:HPV', 
        'STDs: Number of diagnosis', 
        'Dx:Cancer', 
        'Dx:CIN', 
        'Dx:HPV', 
        'Dx']]
y = df['Biopsy']
#excluding 2 columns as features, time since first and last diagnosis of STDs

In [25]:
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42)

In [26]:
train_nulls = X_train.loc[X_train['STDs'].isna()]

In [27]:
train_nulls

Age Number of sexual partners First sexual intercourse  ... Dx:CIN Dx:HPV Dx
417   19                       NaN                     17.0  ...      0      0  0
696   19                       2.0                     16.0  ...      0      0  0
772   18                       7.0                     14.0  ...      0      0  0
90    36                       1.0                     13.0  ...      0      0  0
705   44                       3.0                     23.0  ...      0      0  0
..   ...                       ...                      ...  ...    ...    ... ..
769   26                       2.0                     15.0  ...      1      0  1
330   21                       1.0                     15.0  ...      0      0  0
700   28                       1.0                     17.0  ...      0      0  0
71    34                       NaN                      NaN  ...      0      0  0
102   31                       3.0                     18.0  ...      0      0  0

[86 rows x 30 columns]

In [28]:
X_train_dropped = X_train.drop(train_nulls.index)
#dropping all of those who did not respond to the STD portion of the questionnaire.

In [29]:
mean_impute = SimpleImputer(strategy='mean')
mode_impute = SimpleImputer(strategy='most_frequent')

In [30]:
X_train_dropped[['Smokes']] = mode_impute.fit_transform(X_train_dropped[['Smokes']])
#imputing mode, as this will either be smoking or non smoking, a binary option
#this will cause some error, as some people will be non smokers that this assumes are smokers, or vice versa

In [31]:
X_train_dropped[['Smokes (years)', 'Smokes (packs/year)']] = mean_impute.fit_transform(X_train_dropped[['Smokes (years)', 'Smokes (packs/year)']])
#imputing mean, which will be above 0 (affected by years smoking and packs/year of smokers)
#but also affected by 0's (people who smoked 0 years and packs/year, non-smokers)
#this helps to neutralize the error caused in the previous step

In [32]:
X_train_dropped[['Hormonal Contraceptives']] = mode_impute.fit_transform(X_train_dropped[['Hormonal Contraceptives']])
X_train_dropped[['Hormonal Contraceptives (years)']] = mean_impute.fit_transform(X_train_dropped[['Hormonal Contraceptives (years)']])
#reasoning same as above 2 cells

In [33]:
X_train_dropped[['IUD']] = mode_impute.fit_transform(X_train_dropped[['IUD']])
X_train_dropped[['IUD (years)']] = mean_impute.fit_transform(X_train_dropped[['IUD (years)']])
#reasoning same as above 3 cells

In [34]:
X_train_dropped[['Number of sexual partners', 'Num of pregnancies', 'First sexual intercourse']] = mode_impute.fit_transform(X_train_dropped[['Number of sexual partners', 'Num of pregnancies', 'First sexual intercourse']])
#imputing mode to avoid strong effects of outliers

In [35]:
X_train_dropped.isna().sum()

Age                                   0
Number of sexual partners             0
First sexual intercourse              0
Num of pregnancies                    0
Smokes                                0
Smokes (years)                        0
Smokes (packs/year)                   0
Hormonal Contraceptives               0
Hormonal Contraceptives (years)       0
IUD                                   0
IUD (years)                           0
STDs                                  0
STDs (number)                         0
STDs:condylomatosis                   0
STDs:cervical condylomatosis          0
STDs:vaginal condylomatosis           0
STDs:vulvo-perineal condylomatosis    0
STDs:syphilis                         0
STDs:pelvic inflammatory disease      0
STDs:genital herpes                   0
STDs:molluscum contagiosum            0
STDs:AIDS                             0
STDs:HIV                              0
STDs:Hepatitis B                      0
STDs:HPV                              0
